### DISCUSSION TITLE
how to use time varying universe？

### DISCUSSION THREAD

**Discussion by temph2020** [cvxgrp/cvxportfolio#183] [Q&A]
for example，use top 100  biggest volume stocks as universe

**Comment by enzbus** [comment]
You should prepare your returns and volumes (optional) data frames, pass them to UserProvidedMarketData and specify universe_selection_in_time for your time varying universe selection. Documentation is here https://www.cvxportfolio.com/en/stable/data.html#market-data-servers . Once you do that things should work in the same way as in the examples with market data downloaded internally. In general things are more complicated when you provide your own market data since it may not be formatted as cleanly (missing values, ...). It is a work in progress item to give more examples of usage with custom data, but if you try and run into problems feel free to open issues!
